microbench/db_basic_bench.cc

```diff
@@ -129,6 +245,1027 @@ static void DBClose(benchmark::State& state) {
+    if (enable_statistics) {
+      HistogramData histogram_data;
+      options.statistics->histogramData(COMPACTION_TIME, &histogram_data);
+      state.counters["comp_time"] = histogram_data.average;
+      options.statistics->histogramData(COMPACTION_CPU_TIME, &histogram_data);
+      state.counters["comp_cpu_time"] = histogram_data.average;
+      options.statistics->histogramData(COMPACTION_OUTFILE_SYNC_MICROS,
+                                        &histogram_data);
+      state.counters["comp_outfile_sync"] = histogram_data.average;
+
+      state.counters["comp_read"] = static_cast<double>(
+          options.statistics->getTickerCount(COMPACT_READ_BYTES));
+      state.counters["comp_write"] = static_cast<double>(
+          options.statistics->getTickerCount(COMPACT_WRITE_BYTES));
+
+      state.counters["user_key_comparison_count"] =
+          static_cast<double>(get_perf_context()->user_key_comparison_count);
+      state.counters["block_read_count"] =
+          static_cast<double>(get_perf_context()->block_read_count);
+      state.counters["block_read_time"] =
+          static_cast<double>(get_perf_context()->block_read_time);
+      state.counters["block_checksum_time"] =
+          static_cast<double>(get_perf_context()->block_checksum_time);
+      state.counters["new_table_block_iter_nanos"] =
+          static_cast<double>(get_perf_context()->new_table_block_iter_nanos);
+      state.counters["new_table_iterator_nanos"] =
+          static_cast<double>(get_perf_context()->new_table_iterator_nanos);
+      state.counters["find_table_nanos"] =
+          static_cast<double>(get_perf_context()->find_table_nanos);
+    }
+
+    TeardownDB(state, db, options, kg);
+  }
+}
+
+static void ManualCompactionArguments(benchmark::internal::Benchmark* b) {
+  for (int comp_style : {kCompactionStyleLevel, kCompactionStyleUniversal}) {
+    for (int64_t max_data : {32l << 20, 128l << 20}) {
+      for (int64_t per_key_size : {256, 1024}) {
+        for (bool enable_statistics : {false, true}) {
+          b->Args({comp_style, max_data, per_key_size, enable_statistics});
+        }
+      }
+    }
+  }
+  b->ArgNames({"comp_style", "max_data", "per_key_size", "enable_statistics"});
+}
+
+BENCHMARK(ManualCompaction)->Iterations(1)->Apply(ManualCompactionArguments);
+
+static void ManualFlush(benchmark::State& state) {
+  uint64_t key_num = state.range(0);
+  uint64_t per_key_size = state.range(1);
+  bool enable_statistics = true;
+
+  // setup DB
+  static DB* db;
+  Options options;
+  if (enable_statistics) {
+    options.statistics = CreateDBStatistics();
+  }
+  options.disable_auto_compactions = true;
+  options.level0_file_num_compaction_trigger = (1 << 30);
+  options.level0_slowdown_writes_trigger = (1 << 30);
+  options.level0_stop_writes_trigger = (1 << 30);
+  options.soft_pending_compaction_bytes_limit = 0;
+  options.hard_pending_compaction_bytes_limit = 0;
+  options.write_buffer_size = 2 << 30;  // 2G to avoid auto flush
```

Comment:
Maybe a good idea to cast 2 first before shifting.

Reply:
sorry, do you mean like:
```
options.write_buffer_size = 2l << 30;
```

Reply:
`2ull << 30` or `static_cast<size_t>(2) << 30`